future/atomicTransactionComposer.go

```diff
@@ -0,0 +1,540 @@
+ *   corresponds to the signed transaction from txnGroup[indexesToSign[i]]
+ */
+type TransactionSigner = func(txGroup []types.Transaction, indexesToSign []int) ([][]byte, []string, error)
+
+/**
+ * Create a TransactionSigner that can sign transactions for the provided basic Account.
+ */
+func MakeBasicAccountTransactionSigner(account crypto.Account) TransactionSigner {
+
+	return func(txGroup []types.Transaction, indexesToSign []int) ([][]byte, []string, error) {
+		var stxs [][]byte
+		var txids []string
+		for _, pos := range indexesToSign {
+			txid, stxBytes, err := crypto.SignTransaction(account.PrivateKey, txGroup[pos])
+			if err != nil {
+				return nil, nil, err
+			}
+
+			stxs = append(stxs, stxBytes)
+			txids = append(txids, txid)
+		}
+
+		return stxs, txids, nil
+	}
+}
+
+/**
+ * Create a TransactionSigner that can sign transactions for the provided LogicSigAccount.
+ */
+func MakeLogicSigAccountTransactionSigner(logicSigAccount crypto.LogicSigAccount) TransactionSigner {
+
+	return func(txGroup []types.Transaction, indexesToSign []int) ([][]byte, []string, error) {
+		var stxs [][]byte
+		var txids []string
+		for _, pos := range indexesToSign {
+			txid, stxBytes, err := crypto.SignLogicSigAccountTransaction(logicSigAccount, txGroup[pos])
+			if err != nil {
+				return nil, nil, err
+			}
+
+			stxs = append(stxs, stxBytes)
+			txids = append(txids, txid)
+		}
+
+		return stxs, txids, nil
+	}
+}
+
+/**
+ * Create a TransactionSigner that can sign transactions for the provided Multisig account.
+ * @param msig - The Multisig account metadata
+ * @param sks - An array of private keys belonging to the msig which should sign the transactions.
+ */
+func MakeMultiSigAccountTransactionSigner(msig crypto.MultisigAccount, sks [][]byte) TransactionSigner {
+
+	return func(txGroup []types.Transaction, indexesToSign []int) ([][]byte, []string, error) {
+		var stxs [][]byte
+		var txids []string
+		for _, pos := range indexesToSign {
+			var unmergedStxs [][]byte
+			var txid string
+			for _, sk := range sks {
+				tempTxid, unmergedStxBytes, err := crypto.SignMultisigTransaction(sk, msig, txGroup[pos])
+				txid = tempTxid
+				if err != nil {
+					return nil, nil, err
+				}
+
+				unmergedStxs = append(unmergedStxs, unmergedStxBytes)
+			}
+
+			if len(sks) > 1 {
+				tempTxid, stxBytes, err := crypto.MergeMultisigTransactions(unmergedStxs...)
+				txid = tempTxid
+				if err != nil {
+					return nil, nil, err
+				}
+
+				stxs = append(stxs, stxBytes)
+				txids = append(txids, txid)
+			} else {
+				stxs = append(stxs, unmergedStxs[0])
+				txids = append(txids, txid)
+			}
+		}
+
+		return stxs, txids, nil
+	}
+}
+
+/** Represents an unsigned transactions and a signer that can authorize that transaction. */
+type TransactionWithSigner struct {
+	/** An unsigned transaction */
+	Txn types.Transaction
+	/** A transaction signer that can authorize txn */
+	Signer TransactionSigner
+}
+
+type MethodArgument interface {
```

Comment:
What's the purpose of this interface and the `ABIValue` type? It seems like it just adds more friction to calling a method, since if I want to call `add(uint32,uint32)uint64` I have to make the args array like this:

```go
// get the ABI type, which is redundant since it's already in the Method object
uint32Type, err := abi.TypeOf("uint32")

args := []MethodArgument{
	ABIValue{
		AbiType:  uint32Type,
		RawValue: uint32(1),
	},
	ABIValue{
		AbiType:  uint32Type,
		RawValue: uint32(2),
	},
}
```

Ideally the caller could just do `args := []interface{}{ uint32(1), uint32(2) }`. I guess you need to get the `abi.Type` from somewhere, so maybe a method on `Arg` could be added to parse and return this?

Reply:
Ah I see that you may have copied the `MethodArgument` and `ABIValue` types from the spec doc. For this SDK they likely don't make sense for the reasons I mentioned about.